File: src/main/java/me/kansio/client/modules/impl/movement/flight/misc/Funcraft.java
package me.kansio.client.modules.impl.movement.flight.misc;

import me.kansio.client.event.impl.BlockCollisionEvent;
import me.kansio.client.event.impl.MoveEvent;
import me.kansio.client.event.impl.PacketEvent;
import me.kansio.client.event.impl.UpdateEvent;
import me.kansio.client.modules.impl.movement.flight.FlightMode;
import me.kansio.client.utils.math.MathUtil;
import me.kansio.client.utils.math.Stopwatch;
import me.kansio.client.utils.player.PlayerUtil;
import net.minecraft.potion.Potion;

public class Funcraft extends FlightMode {

    private double moveSpeed, lastDist, ascension;
    private int level, wait, xd, xdnewtest;
    private Stopwatch timer = new Stopwatch();

    public Funcraft() {
        super("Funcraft");
    }

    @Override
    public void onUpdate(UpdateEvent event) {
        if (!event.isPre()) {
            double xDist = mc.thePlayer.posX - mc.thePlayer.prevPosX;
            double zDist = mc.thePlayer.posZ - mc.thePlayer.prevPosZ;
            lastDist = Math.sqrt((xDist * xDist) + (zDist * zDist));
        }

        if (event.isPre()) {
            if (mc.gameSettings.keyBindJump.isKeyDown()) {
                mc.thePlayer.setPosition(mc.thePlayer.posX, mc.thePlayer.posY + 0.15, mc.thePlayer.posZ);
                mc.thePlayer.motionY = 0.15;
            } else if (mc.gameSettings.keyBindSneak.isKeyDown()) {
                mc.thePlayer.setPosition(mc.thePlayer.posX, mc.thePlayer.posY - 0.15, mc.thePlayer.posZ);
                mc.thePlayer.motionY = -0.15;
            } else mc.thePlayer.motionY = 0;
            if (mc.getCurrentServerData() != null && mc.getCurrentServerData().serverIP != null && mc.getCurrentServerData().serverIP.toLowerCase().contains("hypixel")) {
                event.setOnGround(true);
            }
            double result = 0.00000000334947 + MathUtil.getRandomInRange(0.00000000014947, 0.00000000064947);
            if (mc.thePlayer.ticksExisted % 3 == 0) {
                event.setPosY(mc.thePlayer.posY + result);
                event.setOnGround(false);
            }
                        /*if (mc.thePlayer.ticksExisted % 3 == 0) {
                            ascension += value;
                        }
                        if (ascension > value * 7) {
                            event.setOnGround(false);
                            event.setY(mc.thePlayer.posY + ascension);
                            ascension /= 1.125F;
                        }*/
            // Printer.print(""+event.getY());
            if ((mc.thePlayer.moveForward != 0.0F || mc.thePlayer.moveStrafing != 0.0F) && mc.thePlayer.onGround) {
                if (!getFlight().getBoost().getValue()) {
                    final float motionY = 0.42f + (mc.thePlayer.isPotionActive(Potion.jump) ? ((mc.thePlayer.getActivePotionEffect(Potion.jump).getAmplifier() + 1) * 0.1F) : 0);
                    //mc.thePlayer.motionY = motionY;
                    //   mc.thePlayer.setPosition(mc.thePlayer.posX, mc.thePlayer.posY + motionY, mc.thePlayer.posZ);
                }
                if ((getFlight().getBoostMode().getValue().equals("Damage") || getFlight().getBoostMode().getValue().equals("WOWOMG")) && getFlight().getBoost().getValue())
                    if (xd == 0) {
                       PlayerUtil.damagePlayer();
                        xd = 1;
                    }
            }
        }
    }

    @Override
    public void onMove(MoveEvent event) {
        if (getFlight().getExtraBoost().getValue() && getFlight().getBoost().getValue()) {
            if (!timer.timeElapsed(135) && timer.timeElapsed(20)) {
                mc.timer.timerSpeed = 3.5f;
            } else {
                mc.timer.timerSpeed = 1.0f;
            }
            if (level < 20) {
                //TimerUtil.Timer(3.6f);
                timer.resetTime();
            }
        }
        if (getFlight().getBoost().getValue()) {
            final float motionY = 0.42f + (mc.thePlayer.isPotionActive(Potion.jump) ? ((mc.thePlayer.getActivePotionEffect(Potion.jump).getAmplifier() + 1) * 0.1F) : 0);
            switch (getFlight().getBoostMode().getValue()) {
                case "Normal":
                case "Damage":
                    if (mc.thePlayer.isMoving()) {
                        if (level != 1) {
                            if (level == 2) {
                                //mc.timer.timerSpeed = Math.max(1, 3.5F);
                                ++level;
                                moveSpeed *= mc.thePlayer.isPotionActive(Potion.moveSpeed) ? getFlight().getSpeed().getValue() - 0.3 : getFlight().getSpeed().getValue();
                                //Printer.print("2: "+moveSpeed);
                            } else if (level == 3) {
                                ++level;
                                double difference = 0.1 * (lastDist - getBaseMoveSpeed());
                                moveSpeed = lastDist - difference;
                            } else {
                                level++;
                                if (mc.theWorld.getCollidingBoundingBoxes(mc.thePlayer, mc.thePlayer.getEntityBoundingBox().offset(0.0D, mc.thePlayer.motionY, 0.0D)).size() > 0 || mc.thePlayer.isCollidedVertically) {
                                    level = 1;
                                }
                                moveSpeed = lastDist - lastDist / 159D;
                            }
                        } else if (mc.thePlayer.hurtResistantTime == 19 || getFlight().getBoostMode().getValue().equals("Normal")) {
                            event.setMotionY(mc.thePlayer.motionY = motionY);
                            level = 2;
                            double boost = mc.thePlayer.isPotionActive(Potion.moveSpeed) ? 1.5 : 1.62;
                            moveSpeed = boost * getBaseMoveSpeed() - 0.01D;
                        }
                    } else {
                        moveSpeed = 0;
                    }
                    //Printer.print("f: "+moveSpeed + " " + level);
                    moveSpeed = Math.max(moveSpeed, getBaseMoveSpeed());
                    if (level == 1 && mc.thePlayer.hurtResistantTime != 19 && !getFlight().getBoostMode().getValue().equals("Normal")) moveSpeed = 0.011;
                    PlayerUtil.setMotion(event, moveSpeed);
                    break;
                case "WOWOMG":
                    if (mc.thePlayer.isMoving()) {
                        if (level != 1) {
                            if (level == 2) {
                                //mc.timer.timerSpeed = Math.max(1, 3.5F);
                                ++level;
                                moveSpeed *= mc.thePlayer.isPotionActive(Potion.moveSpeed) ? getFlight().getSpeed().getValue() - 0.3 : getFlight().getSpeed().getValue();
                                //Printer.print("2: "+moveSpeed);
                            } else if (level == 3) {
                                ++level;
                                double difference = 0.01 * (lastDist - getBaseMoveSpeed());
                                moveSpeed = lastDist - difference;
                            } else {
                                level++;
                                if (mc.theWorld.getCollidingBoundingBoxes(mc.thePlayer, mc.thePlayer.getEntityBoundingBox().offset(0.0D, mc.thePlayer.motionY, 0.0D)).size() > 0 || mc.thePlayer.isCollidedVertically) {
                                    level = 1;
                                }
                                moveSpeed = moveSpeed - moveSpeed / 159.9D;
                            }
                        } else {
                            event.setMotionY(mc.thePlayer.motionY = motionY);
                            level = 2;
                            double boost = mc.thePlayer.isPotionActive(Potion.moveSpeed) ? 1.55 : 1.62;
                            moveSpeed = boost * getBaseMoveSpeed() - 0.01D;
                        }
                    } else {
                        moveSpeed = 0;
                    }
                    //  Printer.print("f: "+moveSpeed + " " + level);
                    moveSpeed = Math.max(moveSpeed, getBaseMoveSpeed());
                    PlayerUtil.setMotion(event, moveSpeed);
                    break;
            }
        } else {
            PlayerUtil.setMotion(event, getBaseMoveSpeed());
            //TimerUtil.Timer(0.5f);
            //MoveUtil.TP(event, 0.28, 0);
        }
    }

    private double getBaseMoveSpeed() {
        double n = 0.2873;
        if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) {
            n *= 1.0 + 0.2 * (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1);
        }
        return n;
    }

    @Override
    public void onPacket(PacketEvent event) {
        super.onPacket(event);
    }

    @Override
    public void onCollide(BlockCollisionEvent event) {
        super.onCollide(event);
    }

    @Override
    public void onEnable() {
        super.onEnable();
    }

    @Override
    public void onDisable() {
        lastDist = 0.0D;
        xdnewtest = 0;
        ascension = 0;
        xd = 0;
    }
}

File: src/main/java/today/sleek/client/modules/impl/movement/flight/misc/Funcraft.kt
package today.sleek.client.modules.impl.movement.flight.misc

import today.sleek.base.event.impl.MoveEvent
import today.sleek.base.event.impl.UpdateEvent
import today.sleek.client.modules.impl.movement.flight.FlightMode
import today.sleek.client.utils.player.PlayerUtil
import net.minecraft.potion.Potion
import kotlin.math.sqrt

// speed -= speed / 152
class Funcraft : FlightMode("Funcraft") {
    var thing = 0
    var speed = 2.5
    var boosted = false
    var lastDist = 0.0;

    override fun onUpdate(event: UpdateEvent?) {
        if (!event!!.isPre) {
            lastDist = sqrt((mc.thePlayer.posX - mc.thePlayer.prevPosX) * (mc.thePlayer.posX - mc.thePlayer.prevPosX) + (mc.thePlayer.posZ - mc.thePlayer.prevPosZ) * (mc.thePlayer.posZ - mc.thePlayer.prevPosZ))
        }
        mc.thePlayer.motionY = 0.0.also {
            if (thing > 1) {
                mc.thePlayer.setPosition(
                    mc.thePlayer.posX, mc.thePlayer.posY - 8E-6, mc.thePlayer.posZ
                )
            }
        }
        if (mc.thePlayer.isMoving && !mc.thePlayer.isCollidedHorizontally) {
            if (boosted || speed > 0.25) {
//                speed -= speed / 152
                mc.timer.timerSpeed = 1.8f

            }
            if (speed < 0.36) {
                mc.timer.timerSpeed = 1.34f

            }
            if (!boosted) {
                speed = 0.0
            }
        } else {
            speed = 0.0
        }
    }

    override fun onMove(event: MoveEvent?) {


//        ChatUtil.log("Before: $thing | $speed")
        ++thing
        when (thing) {
            1 -> {

                boosted = true
                event!!.motionY = 0.42.also { mc.thePlayer.motionY = it }
                val boost = if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) 1.65 else 1.78
                speed = boost * PlayerUtil.getBaseSpeed()

            }
            2 -> {
                speed *= flight.speed.value
            }
            3 -> {
                val diff = 0.1 * (lastDist - PlayerUtil.getBaseSpeed())
                speed = lastDist - diff
            }
            else -> {//
                if (mc.thePlayer.isCollidedVertically || mc.theWorld.getCollidingBoundingBoxes(
                        mc.thePlayer, mc.thePlayer.entityBoundingBox.offset(0.0, mc.thePlayer.motionY, 0.0)
                    ).size > 0
                )
                    thing = 1;

                speed -= speed / 152
            }
        }
//        ChatUtil.log("After: $thing | $speed")
//        event!!.motionY = 0.0; also {
//            mc.thePlayer.motionY = 0.0
//            mc.thePlayer.setPosition(
//                mc.thePlayer.posX, mc.thePlayer.posY - 8E-6, mc.thePlayer.posZ
//            )
////            ChatUtil.log("" + (mc.thePlayer.posY - 8E-6))
//        }
//        if (thingy == 2) {
//            speed *= if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) flight.speed.value - 0.3 else flight.speed.value
//            thingy = 3
//        }
//        if (mc.thePlayer.isMoving) {
//            if (mc.thePlayer.onGround) {
//                event.motionY = 0.42; also {
//                    mc.thePlayer.motionY = 0.42
//                    speed = 0.8
//                }
//                boosted = true
//
//            }
//        }
        // PlayerUtil.getBaseSpeed().toDouble().coerceAtLeast(speed) = Math.max(PlayerUtil.getBaseSpeed(), speed)
        PlayerUtil.setMotion(event, PlayerUtil.getBaseSpeed().toDouble().coerceAtLeast(speed))


    }

    override fun onEnable() {
        boosted = false
        lastDist = 0.0
        thing = 0
    }
}